File: IQ/Classes/clsUnit.vb
Imports dataAccess

Public Class clsUnit
    Implements i_Editable

    Property ID As Integer
    Property Translation As clsTranslation 'carries the translation.key - and exposes (via an indexed defautl property) the underlying text
    Property Symbol As String
    Property Code As String  'our internal code for referencing these units eg KG (most of the time it will be the same as the name)
    Property MeasureID As Integer

    Dim oCode As String

    Public Function DisplayName(language As clsLanguage) As String Implements i_Editable.displayName
    
        Return Me.Translation.text(language)
    End Function

    Public Sub New()


    End Sub

    Public Sub delete(ByRef errormessages As List(Of String)) Implements i_Editable.delete



        Try
            da.DBExecutesql("DELETE FROM UNIT WHERE ID=" & Me.ID)  'will often fail due to RI (expose this error through the editor)
        Catch ex As Exception
            errormessages.Add(ex.Message.ToString)
        End Try

    End Sub


    Public Function Insert(ByRef errormessages As List(Of String)) As Object Implements i_Editable.Insert

        Return New clsUnit(Me.Code, Me.Translation, Me.Symbol,Me.MeasureID)

    End Function

    Public Sub Update(ByRef errormessages As List(Of String)) Implements i_Editable.update

        Dim sql$
        sql$ = "UPDATE [Unit] set "
        sql$ &= "code=" & da.SqlEncode(Me.Code) & ","
        sql$ &= "fk_translation_key_name=" & Me.Translation.Key & ","
        sql$ &= "symbol=" & da.SqlEncode(Me.Symbol)
        sql$ &= " WHERE ID=" & Me.ID

        iq.i_unit_code.Remove(oCode)
        iq.i_unit_code.Add(Me.Code, Me)

        oCode = Me.Code

        da.DBExecutesql(sql$)

    End Sub


    Public Sub New(ByVal code As String, ByVal translation As clsTranslation, ByVal Symbol As String, MeasureID As Integer)

        Me.Translation = translation
        Me.Symbol = Symbol
        Me.Code = code
        Me.MeasureID = MeasureID

        Dim sql$
        sql$ = "Insert into [Unit] ([code],[FK_Translation_key_name],[symbol],FK_Measure_ID) values (" & da.SqlEncode(code) & "," & translation.Key & "," & da.SqlEncode(Me.Symbol) & "," & da.SqlEncode(Me.MeasureID) & ");"
        Me.ID = da.DBExecutesql(sql$, True)

        iq.Units.Add(Me.ID, Me)
        If Not iq.i_unit_code.ContainsKey(code) Then  'hmm not sure why this is needed 
            iq.i_unit_code.Add(Me.Code, Me)
        End If

        oCode = Me.Code


    End Sub



    Public Sub New(ByVal id As Integer, ByVal code As String, ByVal translation As clsTranslation, ByVal Symbol As String, MeasureID As Integer)

        Me.ID = id
        Me.Translation = translation
        Me.Symbol = Symbol
        Me.Code = code
        Me.MeasureID = MeasureID

        iq.Units.Add(Me.ID, Me)

        iq.i_unit_code.Add(Me.Code, Me)

        oCode = Me.Code

    End Sub

End Class

File: IQ/Classes/clsUnit.cs
using dataAccess;


public class clsUnit : i_Editable
{

    public int ID { get; set; }
    public clsTranslation Translation { get; set; } //carries the translation.key - and exposes (via an indexed defautl property) the underlying text
    public string Symbol { get; set; }
    public string Code { get; set; } //our internal code for referencing these units eg KG (most of the time it will be the same as the name)
    public int MeasureID { get; set; }

    string oCode;

    public string displayName(clsLanguage language)
    {

        return this.Translation.text(language);
    }

    public clsUnit()
    {


    }

    public void delete(ref List<string> errormessages)
    {



        try
        {
            da.DBExecutesql("DELETE FROM UNIT WHERE ID=" + System.Convert.ToString(this.ID)); //will often fail due to RI (expose this error through the editor)
        }
        catch (Exception ex)
        {
            errormessages.Add(ex.Message.ToString());
        }

    }


    public dynamic Insert(ref List<string> errormessages)
    {

        return new clsUnit(this.Code, this.Translation, this.Symbol, this.MeasureID);

    }

    public void update(ref List<string> errormessages)
    {

        object sql = null;
        sql = "UPDATE [Unit] set ";
        sql += "code=" + da.SqlEncode(this.Code) + ",";
        sql += "fk_translation_key_name=" + this.Translation.Key + ",";
        sql += "symbol=" + da.SqlEncode(this.Symbol);
        sql += " WHERE ID=" + System.Convert.ToString(this.ID);

        iq.i_unit_code.Remove(oCode);
        iq.i_unit_code.Add(this.Code, this);

        oCode = this.Code;

        da.DBExecutesql(sql);

    }


    public clsUnit(string code, clsTranslation translation, string Symbol, int MeasureID)
    {

        this.Translation = translation;
        this.Symbol = Symbol;
        this.Code = code;
        this.MeasureID = MeasureID;

        object sql = null;
        sql = "Insert into [Unit] ([code],[FK_Translation_key_name],[symbol],FK_Measure_ID) values (" + da.SqlEncode(code) + "," + Translation.Key + "," + da.SqlEncode(this.Symbol) + "," + da.SqlEncode(this.MeasureID) + ");";
        this.ID = System.Convert.ToInt32(da.DBExecutesql(sql, true));

        iq.Units.Add(this.ID, this);
        if (!iq.i_unit_code.ContainsKey(code)) //hmm not sure why this is needed
        {
            iq.i_unit_code.Add(this.Code, this);
        }

        oCode = this.Code;


    }



    public clsUnit(int id, string code, clsTranslation translation, string Symbol, int MeasureID)
    {

        this.ID = id;
        this.Translation = translation;
        this.Symbol = Symbol;
        this.Code = code;
        this.MeasureID = MeasureID;

        iq.Units.Add(this.ID, this);

        iq.i_unit_code.Add(this.Code, this);

        oCode = this.Code;

    }

}